lib/interface_quda.cpp

```diff
@@ -4515,135 +4522,226 @@ void computeCloverForceQuda(void *h_mom, double dt, void **h_x, void **, double
   fParam.setPrecision(fParam.Precision(), true);
   GaugeField cudaForce(fParam);
 
-  ColorSpinorParam qParam;
-  qParam.location = QUDA_CUDA_FIELD_LOCATION;
-  qParam.nColor = 3;
-  qParam.nSpin = 4;
-  qParam.siteSubset = QUDA_FULL_SITE_SUBSET;
-  qParam.siteOrder = QUDA_EVEN_ODD_SITE_ORDER;
-  qParam.nDim = 4;
+  ColorSpinorParam qParam(nullptr, *inv_param, fParam.x, false, QUDA_CUDA_FIELD_LOCATION);
   qParam.setPrecision(fParam.Precision(), fParam.Precision(), true);
-  qParam.pad = 0;
-  for(int dir=0; dir<4; ++dir) qParam.x[dir] = fParam.x[dir];
-
-  // create the device quark field
   qParam.create = QUDA_NULL_FIELD_CREATE;
   qParam.gammaBasis = QUDA_UKQCD_GAMMA_BASIS;
 
-  std::vector<ColorSpinorField*> quarkX, quarkP;
-  for (int i=0; i<nvector; i++) {
-    quarkX.push_back(ColorSpinorField::Create(qParam));
-    quarkP.push_back(ColorSpinorField::Create(qParam));
-  }
-
-  qParam.siteSubset = QUDA_PARITY_SITE_SUBSET;
-  qParam.x[0] /= 2;
-  ColorSpinorField tmp(qParam);
+  std::vector<ColorSpinorField> x(nvector), p(nvector);
+  for (int i = 0; i < nvector; i++) {
+    p[i] = ColorSpinorField(qParam);
+    x[i] = ColorSpinorField(qParam);
 
-  // create the host quark field
-  qParam.location = QUDA_CPU_FIELD_LOCATION;
-  qParam.create = QUDA_REFERENCE_FIELD_CREATE;
-  qParam.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
-  qParam.gammaBasis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS; // need expose this to interface
+    if (!inv_param->use_resident_solution) {
+      ColorSpinorParam cpuParam(h_x[i], *inv_param, fParam.x, true, inv_param->input_location);
+      ColorSpinorField cpuQuarkX(cpuParam);
+      x[i].Even() = cpuQuarkX;
+      gamma5(x[i].Even(), x[i].Even());
+    } else {
+      x[i].Even() = solutionResident[i];
+    }
+  }
 
-  bool pc_solve = (inv_param->solve_type == QUDA_DIRECT_PC_SOLVE) ||
```

Comment:
Should there still be some type of check that everything passed in is single parity?

Reply:
Not sure what this question is here?  The quark fields that are passed in are all single parity.

Reply:
We discussed this offline, @maddyscientist is going to add a check to make sure the solution type is `matpc` or `normoppc`

Reply:
@maddyscientist thank you for the warning, I enforce `QUDA_MATPCDAG_MATPC_SOLUTION` in tmLQCD before calling the clover force in [9f7a6d5](https://github.com/etmc/tmLQCD/pull/549/commits/9f7a6d5906fe66faee0d5b5b361f538f45c76386). Before [9f7a6d5](https://github.com/etmc/tmLQCD/pull/549/commits/9f7a6d5906fe66faee0d5b5b361f538f45c76386) we were inheriting the `solution_type` from the inversion, which was `QUDA_MATPC_SOLUTION` when using the MG.

Reply:
Great, thanks for the fast update @Marcogarofalo .  I will also add a check that enforces `QudaInvertParam::matpc_type = EVEN_EVEN_ASYM` since the present design implicitly requires this.  Generalizing this to support to both even-even-asym and odd-odd-asym will be part of the update to fuse the two separate clover force interfaces.

One other thing I noticed from looking at your code: I see that you're doing the momentum accumulation on the host and not in QUDA.  With this is mind, you could avoid the zeroing on the host and the download of the field to the gpu by setting `QudaInvertParam::overwrite_mom = 1` (I already applied this optimization to the test code).  Should give a wee speedup.  Alternatively of course, you could just delegate the update to QUDA and delete the accumulation on the host.

Reply:
Thank you for the wee tip, for the moment I made the easiest solution of the two [b6e5a6d](https://github.com/etmc/tmLQCD/pull/549/commits/b6e5a6d8632ae00eb6009b9a3970d8925c69ca4e).
To do the accumulation on QUDA I'll need to rearrange the momentum before and after the QUDA call.

Reply:
> Great, thanks for the fast update @Marcogarofalo . I will also add a check that enforces `QudaInvertParam::matpc_type = EVEN_EVEN_ASYM` since the present design implicitly requires this. Generalizing this to support to both even-even-asym and odd-odd-asym will be part of the update to fuse the two separate clover force interfaces.
> 
@maddyscientist sorry for the confusion in tmLQCD we need `QUDA_MATPC_ODD_ODD_ASYMMETRIC`

Reply:
That makes sense, when I looked at the code deeper I realized this must be the case 🙂.  Thanks for confirming.